x-pack/plugins/endpoint/public/embeddables/resolver/view/submenu.tsx

```diff
@@ -0,0 +1,188 @@
+ */
+export const subMenuAssets = {
+  initialMenuStatus: i18n.translate('xpack.endpoint.resolver.relatedNotRetrieved', {
+    defaultMessage: 'Related Events have not yet been retrieved.',
+  }),
+  menuError: i18n.translate('xpack.endpoint.resolver.relatedRetrievalError', {
+    defaultMessage: 'There was an error retrieving related events.',
+  }),
+  relatedAlerts: {
+    title: i18n.translate('xpack.endpoint.resolver.relatedAlerts', {
+      defaultMessage: 'Related Alerts',
+    }),
+  },
+  relatedEvents: {
+    title: i18n.translate('xpack.endpoint.resolver.relatedEvents', {
+      defaultMessage: 'Events',
+    }),
+  },
+};
+
+interface ResolverSubmenuOption {
+  optionTitle: string;
+  action: () => unknown;
+  prefix?: number | JSX.Element;
+}
+
+export type ResolverSubmenuOptionList =
+  | ResolverSubmenuOption[]
+  | 'waitingForRelatedEventData'
+  | typeof subMenuAssets.initialMenuStatus;
+
+const OptionList = React.memo(
+  ({
+    subMenuOptions,
+    isLoading,
+  }: {
+    subMenuOptions: ResolverSubmenuOptionList;
+    isLoading: boolean;
+  }) => {
+    const selectableOptions =
+      typeof subMenuOptions !== 'object'
+        ? []
+        : subMenuOptions.map((opt: ResolverSubmenuOption): {
+            label: string;
+            prepend?: ReactNode;
+          } => {
+            return opt.prefix
+              ? {
+                  label: opt.optionTitle,
+                  prepend: <span>{opt.prefix}</span>,
+                }
+              : {
+                  label: opt.optionTitle,
+                  prepend: <span />,
+                };
+          });
+    const [options, setOptions] = useState(selectableOptions);
+    return useMemo(
+      () => (
+        <EuiSelectable
+          singleSelection={true}
+          options={options}
+          onChange={newOptions => {
+            setOptions(newOptions);
+          }}
+          listProps={{ showIcons: true, bordered: true }}
+          isLoading={isLoading}
+        >
+          {list => list}
+        </EuiSelectable>
+      ),
+      [isLoading, options]
+    );
+  }
+);
+
+/**
+ * A Submenu to be displayed in one of two forms:
+ *   1) Provided a collection of `optionsWithActions`: it will call `menuAction` then - if and when menuData becomes available - display each item with an optional prefix and call the supplied action for the options when that option is clicked.
+ *   2) Provided `optionsWithActions` is undefined, it will call the supplied `menuAction` when its host button is clicked.
+ */
+export const NodeSubMenu = styled(
+  React.memo(
+    ({
+      menuTitle,
+      menuAction,
+      optionsWithActions,
+      className,
+    }: { menuTitle: string; className?: string; menuAction: () => unknown } & (
+      | {
+          optionsWithActions:
+            | ResolverSubmenuOptionList
+            | typeof subMenuAssets.initialMenuStatus
+            | 'waitingForRelatedEventData';
+        }
+      | { optionsWithActions?: undefined }

```

Comment:
you could probably just put the `?` on the first type with `optionsWithActions`

e.g.

```
 optionsWithActions?:
            | ResolverSubmenuOptionList
            | typeof subMenuAssets.initialMenuStatus
            | 'waitingForRelatedEventData';
        }
```

As a react component prop, theres probably no point in allowing `<NodeSubMenu optionsWithActions={undefined} />`

but if you intend to support this, you could write 
```
 optionsWithActions?:
            | ResolverSubmenuOptionList
            | typeof subMenuAssets.initialMenuStatus
            | 'waitingForRelatedEventData'
            | undefined
        }
```

Also, the type `typeof subMenuAssets.initialMenuStatus` should be `string`, in which case you can just write

```
 optionsWithActions?:
            | ResolverSubmenuOptionList
            | string
            | undefined
        }
```

Since a type that unions `string` with a string literal will evaluate to `string`, the `'waitingForRelatedEventData'` literal should have no effect on the type.

Here's what i mean about the union-with-a-string-literal comment:
![image](https://user-images.githubusercontent.com/35559/82231145-ba1e1100-98fa-11ea-8323-0b6e7c28ac48.png)

Reply:
I thought that would make it easier for the reader to see the difference between the two ways it renders, but I changed to your rec here: https://github.com/elastic/kibana/pull/64121/commits/1985a31e17cdbdacd7a161c8f152d704f79e311c

Reply:
:straight_ruler: Don't: Try to enumerate different modes of input through type annotations. Do: Make type annotations as basic as possible.